File: C++/212.word-search-ii.cpp
/*
 * @lc app=leetcode id=212 lang=cpp
 *
 * [212] Word Search II
 *
 * https://leetcode.com/problems/word-search-ii/description/
 *
 * algorithms
 * Hard (31.82%)
 * Likes:    1767
 * Dislikes: 89
 * Total Accepted:    160.5K
 * Total Submissions: 504.4K
 * Testcase Example:  '[["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]]\n' +
  '["oath","pea","eat","rain"]'
 *
 * Given a 2D board and a list of words from the dictionary, find all words in
 * the board.
 * 
 * Each word must be constructed from letters of sequentially adjacent cell,
 * where "adjacent" cells are those horizontally or vertically neighboring. The
 * same letter cell may not be used more than once in a word.
 * 
 * 
 * 
 * Example:
 * 
 * 
 * Input: 
 * board = [
 * ⁠ ['o','a','a','n'],
 * ⁠ ['e','t','a','e'],
 * ⁠ ['i','h','k','r'],
 * ⁠ ['i','f','l','v']
 * ]
 * words = ["oath","pea","eat","rain"]
 * 
 * Output: ["eat","oath"]
 * 
 * 
 * 
 * 
 * Note:
 * 
 * 
 * All inputs are consist of lowercase letters a-z.
 * The values of words are distinct.
 * 
 * 
 */

// @lc code=start

struct TrieNode {
    TrieNode *child[26];
    string str;
    TrieNode() : str("") {
        for (auto &x: child) x = nullptr;
    }
};


class Solution {
public:
    
    TrieNode* root = new TrieNode();
    
    void insert(string word) {
        TrieNode *cur = root;
        for (char ch: word) {
            if (!cur->child[ch-'a']) cur->child[ch-'a'] = new TrieNode();
            cur = cur->child[ch-'a'];
        }
        cur->str = word;
    }
    
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        vector<string> ans;
        if (words.empty() || board.empty() || board[0].empty()) return ans;
        
        int n = board.size(), m = board[0].size();
        vector<vector<bool>> visited(n, vector<bool>(m, false));
        
        for (auto &w : words) insert(w);
        
        for (int i=0; i<n; ++i) {
            for (int j=0; j<m; ++j) {
                if (root->child[board[i][j]-'a']) {
                    search(board, root->child[board[i][j]-'a'], i, j, visited, ans);
                }
            }
        }
        return ans;
    }
    
    vector<vector<int>> dirs{{0, -1}, {0, 1}, {-1, 0}, {1, 0}};
    void search(vector<vector<char>>& board, TrieNode* cur, int i, int j, vector<vector<bool>>& visited, vector<string>& ans) { 
        if (!cur->str.empty()) {
            ans.push_back(cur->str);
            cur->str.clear();
        }

        int n = board.size(), m = board[0].size();
        
        visited[i][j] = true;
        for (auto &d : dirs) {
            int x = i+d[0], y = j+d[1];
            
            if (x<0 || x>=n || y<0 || y>=m) continue;
            if (visited[x][y]) continue;
            if (!cur->child[board[x][y]-'a']) continue;
            
            search(board, cur->child[board[x][y]-'a'], x, y, visited, ans);
        }
        visited[i][j] = false;
    }
};
// @lc code=end

File: Python/212.word-search-ii.py
#
# @lc app=leetcode id=212 lang=python
#
# [212] Word Search II
#
# https://leetcode.com/problems/word-search-ii/description/
#
# algorithms
# Hard (30.68%)
# Likes:    1548
# Dislikes: 84
# Total Accepted:    144K
# Total Submissions: 466.3K
# Testcase Example:  '[["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]]\n' +
#   '["oath","pea","eat","rain"]'
#
# Given a 2D board and a list of words from the dictionary, find all words in
# the board.
# 
# Each word must be constructed from letters of sequentially adjacent cell,
# where "adjacent" cells are those horizontally or vertically neighboring. The
# same letter cell may not be used more than once in a word.
# 
# 
# 
# Example:
# 
# 
# Input: 
# board = [
# ⁠ ['o','a','a','n'],
# ⁠ ['e','t','a','e'],
# ⁠ ['i','h','k','r'],
# ⁠ ['i','f','l','v']
# ]
# words = ["oath","pea","eat","rain"]
# 
# Output: ["eat","oath"]
# 
# 
# 
# 
# Note:
# 
# 
# All inputs are consist of lowercase letters a-z.
# The values of words are distinct.
# 
# 
#

# @lc code=start
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isWord = False

class Trie():
    def __init__(self):
        self.root = TrieNode()
    def insert(self, word):
        cur = self.root
        for ch in word:
            if ch not in cur.children:
                cur.children[ch] = TrieNode()
            cur = cur.children[ch]
        cur.isWord = True
    def search(self, word):
        cur = self.root
        for ch in word:
            if ch not in cur.children:
                return False
            cur = cur.chidren[ch]
        return cur.isWord

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        trie = Trie()
        for w in words:
            trie.insert(w)
        
        ans = []
        for i in range(len(board)):
            for j in range(len(board[0])):
                self.dfs(board, i, j, trie.root, "", ans)
        return ans

    def dfs(self, board, i, j, node, path, ans):
        if node.isWord:
            ans.append(path)
            node.isWord = False
        
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]):
            return 
        
        ch = board[i][j]
        if ch not in node.children:
            return
        node = node.children[ch]
        board[i][j] = '#'
        self.dfs(board, i+1, j, node, path+ch, ans)
        self.dfs(board, i-1, j, node, path+ch, ans)
        self.dfs(board, i, j+1, node, path+ch, ans)
        self.dfs(board, i, j-1, node, path+ch, ans)
        board[i][j] = ch 
# @lc code=end

